### ISSUE TITLE
Git via agent fails behind a corporate proxy server

### ISSUE THREAD

**Issue by yugmac** [Microsoft/vsts-agent#670]
## **Agent version and platform**
Version of your agent? 2.109.0
<br>
OS of the machine running the agent? Windows 7 Pro SP1.
<br>
## **VSTS type and version**
VisualStudio.com
<br>
If VisualStudio.com, what is your account name? http://vsosse.visualstudio.com
<br>
## **What's not working?** 
The build agent is running on a corporate network and behind a proxy server. Using an Environment Variable VSTS_HTTP_PROXY set with the proxy server address, the agent registers fine with our visualstudio.com account using a PAT. However when we kick off a build it fails on Get sources trying to download from our Git repo. Error shown below in screenshot:
 
![image](https://cloud.githubusercontent.com/assets/23236859/19965450/eb183114-a1be-11e6-88ca-83858a0ff675.png)

I've tried with adding the proxy creds as environment variables too and can see they get added to the http.proxy value in format username:password@proxyserver however to no avail. 

Note* Git does work on the machine running the agent when using:  
git config --global http.proxy http://mydomain\\myusername:mypassword@myproxyserver:80

It asks for the PAT to authenticate with visualstudio.com. However the above setting allows it to communicate via the proxy server.   

<br>

**Comment by TingluoHuang** [comment]
@yugmac https://www.visualstudio.com/en-us/docs/build/admin/agents/v2-windows#run-the-agent-behind-a-web-proxy
in addition to VSTS_HTTP_PROXY, also set VSTS_HTTP_PROXY_USERNAME and VSTS_HTTP_PROXY_PASSWORD follow the Doc

**Comment by yugmac** [comment]
@TingluoHuang  Hello, I have followed the Doc for setting up the agent behind a proxy server. As I stated in the above issue I have tried to use the username and password variables too however to no avail. 
Using VSTS_HTTP_PROXY_USERNAME and VSTS_HTTP_PROXY_PASSWORD appends the values onto the http.proxy url address but this still does not work.

**Comment by yugmac** [comment]
I noticed that in the console the agent outputs text saying "Agent is running behind proxy server: 'http://myproxyserver:80'" and when it fails the error shows "Failed to connect to myproxyserver port 1080: Connection refused" 
Does it matter that one states port 80 and the other port 1080?

Thanks.

**Comment by TingluoHuang** [comment]
Have you try put port info also in VSTS_HTTP_PROXY? like VSTS_HTTP_PROXY = http://localhost:8888

**Comment by yugmac** [comment]
I have the Environment Variables as follows:
VSTS_HTTP_PROXY = http://myproxyserver:80
VSTS_HTTP_PROXY_USERNAME = domain\username
VSTS_HTTP_PROXY_PASSWORD = password

The agent then communicates with visualstudio.com through the proxy server using port 80 and it works. However Git still fails with the above error message stating port 1080.

**Comment by TingluoHuang** [comment]
have you restart the agent process after you set those vars? can you send me your build log and agent diag log? tihuang@microsoft.com

**Comment by yugmac** [comment]
I have restarted, removed and reconfigured. Updated to the latest agent, tried on various machines but all to no avail. Also ran the agent in interactive mode too as well as a windows service. 
I will email you the log info .

Thanks

**Comment by TingluoHuang** [comment]
also, what kind of Proxy are you using? base on the error, git.exe is trying to use port 1080, which is the default port for SOCKS proxy.

**Comment by yugmac** [comment]
@TingluoHuang  I have sent an email with the logs attached. Also I have requested the information you have asked from our operational security department. I do not have access to the proxy server or any information regarding it.

**Comment by TingluoHuang** [comment]
@yugmac please try https://github.com/Microsoft/vsts-agent/releases/tag/v2.109.2

**Comment by TingluoHuang** [comment]
@yugmac does the new version agent unblock you?

**Comment by yugmac** [comment]
@TingluoHuang sorry for not replying sooner, I was off work on annual leave. Anyway, the new version of the agent works and Git is able to communicate via the proxy server. Thanks! 
However one more thing, is there another way to save proxy credentials rather than using environmental variables? These show the proxy authentication username and password in plain text!
Thanks

**Comment by TingluoHuang** [comment]
@yugmac cool, thanks for your validation. The agent will make sure your credential been masked out in any log file or console output. We haven't find a better way to store your proxy credential xplat, we can use dpapi in windows, about dotnet core haven't port that to linux/osx. 
I will close this issue. Thanks for reporting.

**Comment by yugmac** [comment]
@TingluoHuang  yeah you can close this issue! Thank you. 

Well if you need to use Environment variables, what about not showing them on the VSTS agent dashboard under Capabilities. 

![image](https://cloud.githubusercontent.com/assets/23236859/20352643/db8e0524-ac0f-11e6-97e5-1d655e3656de.png)

Could this be done? I know you mask them in log files and console output however they are still shown in plain text as the screenshot above shows.

**Comment by TingluoHuang** [comment]
@yugmac i will fix it.

**Comment by yugmac** [comment]
@TingluoHuang  Aww brilliant! Thanks! Would you like me to create a new issue for you to work against?

**Comment by TingluoHuang** [comment]
@yugmac sure

**Comment by yugmac** [comment]
@TingluoHuang  Okay new issue #686 thank you! :+1:
